src/Morphir/IR/Type/DataCodec.elm

```diff
@@ -414,6 +449,73 @@ decodeData ir tpe =
+                                                                Value.Apply _ (Value.Constructor _ ( [ [ "morphir" ], [ "s", "d", "k" ] ], [ [ "maybe" ] ], [ "just" ] )) value ->
+                                                                    Decode.succeed value
+
+                                                                -- if the evaluated value "Nothing"
+                                                                Value.Constructor _ ( [ [ "morphir" ], [ "s", "d", "k" ] ], [ [ "maybe" ] ], [ "nothing" ] ) ->
+                                                                    fnName
+                                                                        ++ " returned Nothing"
+                                                                        |> Decode.fail
+
+                                                                -- if the evaluated value is "Ok value"
+                                                                Value.Apply _ (Value.Constructor _ ( [ [ "morphir" ], [ "s", "d", "k" ] ], [ [ "result" ] ], [ "ok" ] )) value ->
+                                                                    Decode.succeed value
+
+                                                                -- if the evaluated value "Err err"
+                                                                Value.Apply _ (Value.Constructor _ ( [ [ "morphir" ], [ "s", "d", "k" ] ], [ [ "result" ] ], [ "err" ] )) err ->
+                                                                    let
+                                                                        _ =
+                                                                            Debug.log "Could not derive value" err
+                                                                    in
+                                                                    fnName
+                                                                        ++ " returned Err"
+                                                                        |> Decode.fail
+
+                                                                -- only Maybe and Result types are expected, we can fail
+                                                                _ ->
+                                                                    "Invalid Return Type for "
+                                                                        ++ fnName
+                                                                        |> Decode.fail
+
+                                                        --Decode.succeed val
+                                                        Err error ->
+                                                            let
+                                                                _ =
+                                                                    Debug.log "Interpreter Evaluation Error: " error
```

Comment:
I assume you left this here by accident.

Reply:
The reason I added this was to support error reporting. Since the type of the error is not known, I couldn't use it as the argument to `Decode.fail`.

Do you think it's unnecessary to have this here?

Reply:
`Decode.fail` takes a `String` message so we only need to turn that error into a String to be able to trickle it up. I don't see a codec for the errors returned by the interpreter though so I understand the difficulty. When this happens you can always fall back to `Debug.toString` which can turn anything into a String. It's in the `Debug` module so not ideal but maybe better then `Debug.log`.